### DISCUSSION TITLE
[@types/mapbox__mapbox-gl-draw] new types added

### DISCUSSION THREAD

**Discussion by joel-daros** [DefinitelyTyped/DefinitelyTyped#64239] [Issues with a @types package]
After this [merged commit](https://github.com/mapbox/mapbox-gl-draw/pull/1100), It would be great to have these new constants and methods exposed in types.

[a comment by an automated account is not shown]

**Comment by Shayan-To** [comment]
@MrSparklle Can you create a PR?

**Comment by joel-daros** [comment]
@Shayan-To Sure I can, but could you give some guidance on best practices. I was thinking in something like this:

To declare [Constants](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/constants.js):
```typescript
declare namespace MapboxDraw {
    // ...
    declare namespace Constants {
       const cursors = {
        ADD: 'add',
        MOVE: 'move',
        DRAG: 'drag',
        POINTER: 'pointer',
        NONE: 'none'
      } as const;
     // ...
    }
```

For Libs methods like [createVertex](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/lib/create_vertex.js):

```typescript
declare namespace MapboxDraw {
    // ...
    const Lib = {
      /**
       * Returns GeoJSON for a Point representing the
       * vertex of another feature.
       *
       * @param {string} parentId
       * @param {Array<number>} coordinates
       * @param {string} path - Dot-separated numbers indicating exactly
       *   where the point exists within its parent feature's coordinates.
       * @param {boolean} selected
       * @return {GeoJSON} Point
       */
      createVertex: (
        parentId: string,
        coordinates: Position,
        path: string,
        selected: boolean
      ): GeoJSON => Point
      // ...
    } as const;
```

**Reply by Shayan-To** [reply]
Please define interfaces, like this (and don't use `const name = value` instead of `const name: Type` in declaration files):

```ts
declare namespace MapboxDraw {
    // ...
    interface Constants {
       readonly cursors: {
        ADD: 'add',
        MOVE: 'move',
        DRAG: 'drag',
        POINTER: 'pointer',
        NONE: 'none'
      };
     // ...
    }

    interface Lib {
      /**
       * Returns GeoJSON for a Point representing the
       * vertex of another feature.
       *
       * @param {string} parentId
       * @param {Array<number>} coordinates
       * @param {string} path - Dot-separated numbers indicating exactly
       *   where the point exists within its parent feature's coordinates.
       * @param {boolean} selected
       * @return {GeoJSON} Point
       */
      createVertex(
        parentId: string,
        coordinates: Position,
        path: string,
        selected: boolean
      ): GeoJSON; // "Point" is the description of the return value.
      // ...
    }
    // ...
}
```

Also, don't forget to add the static fields:

```ts
declare class MapboxDraw implements IControl {
    static modes: MapboxDraw.Modes;
    static constants: MapboxDraw.Constants;
    static lib: MapboxDraw.Lib;
    // ...
}
```

**Reply by joel-daros** [reply]
What the best way to type `ctx` in [featuresAt](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/lib/features_at.js)? `DrawCustomModeThis`?

```javascript
        featuresAt: {
            click: (event: MapMouseEvent, bbox: BBox, ctx: ???) => Feature[];
            touch: (event: MapTouchEvent, bbox: BBox, ctx: ???) => Feature[];
        };
```

**Reply by Shayan-To** [reply]
This seems to be a mistake on `mapbox-gl-draw` side. I opened [an issue](https://github.com/mapbox/mapbox-gl-draw/issues/1156).

I suggest we omit `featuresAt` and add a comment in its place.

**Reply by joel-daros** [reply]
I see. We also have `ctx` parameter in [doubleClickZoom](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/lib/double_click_zoom.js) and [getFeatureAtAndSetCursors](https://github.com/mapbox/mapbox-gl-draw/blob/main/src/lib/get_features_and_set_cursor.js)

The types are almost done, now I'm working on update the tests for the new additions. You can see a preview [here](https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...joel-daros:DefinitelyTyped:master). Fell free to add any comment you understand that is necessary to improve or fix any mistake. I need some help to type `ModeHandler` correctly, as you can see, I temporarily typed as any.

**Reply by Shayan-To** [reply]
I suggest you open a draft PR, so we can discuss things in context and have more GitHub features available to us.

**Reply by Shayan-To** [reply]
(Sorry for the delay, I was on a small trip.)

**Reply by joel-daros** [reply]
> I suggest you open a draft PR, so we can discuss things in context and have more GitHub features available to us.

@Shayan-To did you have time to take a look at my PR? Is there anything else I can do to help?

**Comment by joel-daros** [comment]
Ok. PR sent: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64354
